resteasy-jaxrs/src/main/java/org/jboss/resteasy/core/StringParameterInjector.java

```diff
@@ -80,40 +80,39 @@ protected void initialize(Class type, Type genericType, String paramName, Class
       baseType = type;
       baseGenericType = genericType;
 
-      if (type.isArray()) baseType = type.getComponentType();
-      if (List.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = ArrayList.class;
-      }
-      else if (SortedSet.class.isAssignableFrom(type))
-      {
-         isCollection = true;
-         collectionType = TreeSet.class;
-      }
-      else if (Set.class.isAssignableFrom(type))
+      if (type.isArray())
       {
-         isCollection = true;
-         collectionType = HashSet.class;
+    	  baseType = type.getComponentType();
       }
-      if (isCollection)
+     
+      if (!baseType.isPrimitive())
       {
-         if (genericType != null && genericType instanceof ParameterizedType)
+         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+         if (paramConverter != null) 
          {
-            ParameterizedType zType = (ParameterizedType) genericType;
-            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
-            baseGenericType = zType.getActualTypeArguments()[0];
+        	 return;
          }
-         else
-         {
-            baseType = String.class;
-            baseGenericType = null;
+         
+         collectionType = convertParameterTypeToCollectionType();
+         if(collectionType != null){
+        	 isCollection = true;
+	         if (genericType != null && genericType instanceof ParameterizedType)
+	         {
+	            ParameterizedType zType = (ParameterizedType) baseGenericType;
+	            baseType = Types.getRawType(zType.getActualTypeArguments()[0]);
+	            baseGenericType = zType.getActualTypeArguments()[0];
+	         }
+	         else
+	         {
+	            baseType = String.class;
+	            baseGenericType = null;
+	         }
+	         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
+	         if (paramConverter != null) 
+	         {
+	        	 return;
+	         }
          }
-      }
-      if (!baseType.isPrimitive())
-      {
-         paramConverter = factory.getParamConverter(baseType, baseGenericType, annotations);
-         if (paramConverter != null) return;
 
          unmarshaller = factory.createStringParameterUnmarshaller(baseType);

```

Comment:
re: "But the problem with current code is that in case of a collection it directly jump to point 5, even if user provides a Collection (List Set or SortedSet) with either:"

Yes, I think you're right. It just looks like a mis-reading of the spec.